src/swarm-tui/Swarm/TUI/Model/Event.hs

```diff
@@ -0,0 +1,227 @@
+  | FocusWorldEvent
+  | FocusRobotEvent
+  | FocusREPLEvent
+  | FocusInfoEvent
+  | ToggleCreativeModeEvent
+  | ToggleWorldEditorEvent
+  | ToggleREPLVisibilityEvent
+  deriving (Eq, Ord, Show, Enum, Bounded)
+
+mainEvents :: KeyEvents MainEvent
+mainEvents = allKeyEvents $ \case
+  QuitEvent -> "quit"
+  ViewHelpEvent -> "view help"
+  ViewRobotsEvent -> "view robots"
+  ViewRecipesEvent -> "view recipes"
+  ViewCommandsEvent -> "view commands"
+  ViewMessagesEvent -> "view messages"
+  ViewStructuresEvent -> "view structures"
+  ViewGoalEvent -> "view goal"
+  HideRobotsEvent -> "hide robots"
+  ShowCESKDebugEvent -> "debug CESK"
+  PauseEvent -> "pause"
+  RunSingleTickEvent -> "run single tick"
+  IncreaseTpsEvent -> "increse TPS"
+  DecreaseTpsEvent -> "decrease TPS"
+  FocusWorldEvent -> "focus World"
+  FocusRobotEvent -> "focus Robot"
+  FocusREPLEvent -> "focus REPL"
+  FocusInfoEvent -> "focus Info"
+  ToggleCreativeModeEvent -> "creative mode"
+  ToggleWorldEditorEvent -> "world editor"
+  ToggleREPLVisibilityEvent -> "toggle REPL"
+
+defaultMainBindings :: [(MainEvent, [Binding])]
+defaultMainBindings = allBindings $ \case
+  QuitEvent -> [ctrl 'q']
+  ViewHelpEvent -> [fn 1]
+  ViewRobotsEvent -> [fn 2]
+  ViewRecipesEvent -> [fn 3]
+  ViewCommandsEvent -> [fn 4]
+  ViewMessagesEvent -> [fn 5]
+  ViewStructuresEvent -> [fn 6]
+  ViewGoalEvent -> [ctrl 'g']
+  HideRobotsEvent -> [meta 'h']
+  ShowCESKDebugEvent -> [meta 'd']
+  PauseEvent -> [ctrl 'p']
+  RunSingleTickEvent -> [ctrl 'o']
+  IncreaseTpsEvent -> [ctrl 'x']
+  DecreaseTpsEvent -> [ctrl 'z']
+  FocusWorldEvent -> [meta 'w']
+  FocusRobotEvent -> [meta 'e']
+  FocusREPLEvent -> [meta 'r']
+  FocusInfoEvent -> [meta 't']
+  ToggleCreativeModeEvent -> [ctrl 'v']
+  ToggleWorldEditorEvent -> []
+  ToggleREPLVisibilityEvent -> [meta ',']
+
+-- ----------------------------------------------
+--                 REPL EVENTS
+-- ----------------------------------------------
+
+-- | REPL abstract keybinding events.
+--
+-- See 'Swarm.TUI.Controller.REPLEventHandler'.
+data REPLEvent
+  = CancelRunningProgramEvent
+  | TogglePilotingModeEvent
+  | ToggleCustomKeyHandlingEvent
+  deriving (Eq, Ord, Show, Enum, Bounded)
+
+replEvents :: KeyEvents REPLEvent
+replEvents = allKeyEvents $ \case
+  CancelRunningProgramEvent -> "cancel running program"
+  ToggleCustomKeyHandlingEvent -> "toggle custom key handling"
+  TogglePilotingModeEvent -> "toggle piloting mode"
+
+defaultReplBindings :: [(REPLEvent, [Binding])]
+defaultReplBindings = allBindings $ \case
+  CancelRunningProgramEvent -> [ctrl 'c', bind V.KEsc]
+  TogglePilotingModeEvent -> [meta 'p']
+  ToggleCustomKeyHandlingEvent -> [meta 'k']
+
+-- ----------------------------------------------
+--                 REPL EVENTS
+-- ----------------------------------------------
+
+data WorldEvent
+  = ViewBaseEvent
+  | ShowFpsEvent
+  | MoveViewNorthEvent
+  | MoveViewEastEvent
+  | MoveViewSouthEvent
+  | MoveViewWestEvent
```

Comment:
This can simplify some code that is cardinal-direction-specific:
```suggestion
data WorldEvent
  = ViewBaseEvent
  | ShowFpsEvent
  | MoveViewDirectionEvent AbsoluteDir
```

Reply:
This makes me handwrite `Enum` and `Bounded` and add a few imports, but I admit I wanted to do this too.

The usage becomes much more elegant, I just wish Haskell supported this out of the box.

Reply:
I [asked about this on Discourse](https://discourse.haskell.org/t/enumerating-all-values-of-a-nested-sum-type/9907), and got some good suggestions:
* [`universeGeneric`](https://hackage.haskell.org/package/universe-1.2.3/docs/Data-Universe.html#v:universeGeneric) from the `universe` package
* the [`finitary`](https://hackage.haskell.org/package/finitary-2.1.2.0/docs/Data-Finitary.html) package

Reply:
Thanks, @kostmo; there are some good tips in that discussion. 👍 

I think I was considering `finitary` before, but there was not as much need. Did not know about `universeGeneric`.

I will split that as a separate Refactoring Issue.